### ISSUE TITLE
Different test results on Alpine and Debian (and potentially others; library dependency problem)

### ISSUE THREAD

**Issue by dnsmichi** [Orange-OpenSource/hurl#1082]
## Problem

While looking into GitLab CI/CD integrations, I built my own Docker image using Debian 11 as a base image, and downloaded the DEB package file from the release page. Context in https://gitlab.com/everyonecancontribute/dev/hurl-playground/-/blob/main/Dockerfile and the repository. 

After learning that the entrypoint for the upstream Docker image can be modified within CI/CD, I changed the configuration, expecting to see the same results (all green). Context in https://github.com/Orange-OpenSource/hurl/issues/1077 

Unfortunately, all tests now return `HTTP/1.1` instead of `HTTP/2` and cause the hurl tests to fail. 

- Debian: `https://hn.algolia.com/api/v1/search` is HTTP/2
- Alpine: `https://hn.algolia.com/api/v1/search` is HTTP/1.1

## Analysis

Thanks to @lepapareil for the details in  https://gitlab.com/everyonecancontribute/dev/hurl-playground/-/merge_requests/12#note_1203174103 

The problem is the underlying dependency of libcurl and potential bugs in reporting the correct HTTP protocol version. 

Alpine upstream image

```
$ docker run -ti ghcr.io/orange-opensource/hurl:latest hurl --version

hurl 1.8.0 libcurl/7.86.0-DEV OpenSSL/3.0.7 zlib/1.2.13
Features (libcurl):  alt-svc AsynchDNS HSTS IPv6 libz SSL UnixSockets
Features (built-in): brotli
```

Debian custom image
```
$ docker run -ti registry.gitlab.com/everyonecancontribute/dev/hurl-playground:latest  hurl --version

hurl 1.8.0 libcurl/7.74.0 OpenSSL/1.1.1n zlib/1.2.11 brotli/1.0.9 libidn2/2.3.0 libssh2/1.9.0 nghttp2/1.43.0
Features (libcurl):  alt-svc AsynchDNS brotli HTTP2 IDN IPv6 Largefile libz NTLM NTLM_WB SPNEGO SSL TLS-SRP UnixSockets
Features (built-in): brotli
```

It seems that libcurl on Alpine isn't compiled with HTTP/2 support thus generating these false results. 

## Steps to reproduce 

Hurl test case (local in git or on Gitlab.com using the Web IDE to add), source in https://gitlab.com/everyonecancontribute/dev/hurl-playground/-/blob/main/use-cases/hackernews.hurl 

```
mkdir -p use-cases
cat >use-cases/hackernews.hurl <<EOF
GET https://hn.algolia.com/api/v1/search
[QueryStringParams]
query: gitlab
#query: hurl
tags: front_page

HTTP/2 200

[Asserts]
# Inverted assert: 0 means Ok (no front page), all other values indicate front page alert
jsonpath "$.hits" count == 0

EOF
```

GitLab CI/CD configuration, either using the custom built image, or the upstream image.  Excerpt from https://gitlab.com/everyonecancontribute/dev/hurl-playground/-/blob/c9ce17f8421a5cd761e693bbff49a0ebf95596df/.gitlab-ci.yml

```yaml
# Hurl job template  
.hurl-tmpl:
    #image: registry.gitlab.com/everyonecancontribute/dev/hurl-playground:latest 
    # Use the upstream container image and override the ENTRYPOINT to run CI/CD script
    # https://docs.gitlab.com/ee/ci/docker/using_docker_images.html#override-the-entrypoint-of-an-image 
    image:
        name: ghcr.io/orange-opensource/hurl:latest
        entrypoint: [""]
    variables:
        HURL_URL: gitlab.com
        HURL_JUNIT_REPORT: hurl_junit_report.xml

# Hurl tests from configuration file, generating JUnit report integration in GitLab CI/CD
hurl-report:
    extends: .hurl-tmpl
    script:
      # Hack: Workaround for 'id' instead of 'name' in JUnit report from Hurl. https://gitlab.com/gitlab-org/gitlab/-/issues/299086 and do not fail early,https://stackoverflow.com/a/70087560/1821348 
      - set +e +o pipefail
      - hurl --test use-cases/*.hurl --report-junit $HURL_JUNIT_REPORT; result=$?; sed -i 's/id/name/g' $HURL_JUNIT_REPORT
      - set -e -o pipefail
      - "[ $result == 0 ]"
    artifacts:
      when: always
      paths:
        - $HURL_JUNIT_REPORT
      reports:
        junit: $HURL_JUNIT_REPORT

```

## Proposal 

1. Document the unexpected behaviour in the docs in a troubleshooting section
1. More integration tests that install the package binary on test distributions, and verify the returned tests are ok. Suggest starting with Debian 11, Alpine 3.17 which shows the problem directly.

**Comment by jcamiel** [comment]
I think that, in this case, the system is working as design :

- if you're not interested in HTTP version, your tests can be

```
GET http://foo.com
HTTP/* 200
```

- if you want to be strict on HTTP version (specifically `HTTP/1.0`, `HTTP/1.1` or `HTTP/2`) the failing tests show explicitly that libcurl on Alpine 3.17 doesn't support HTTP2 (as `hurl --version` confirms it).

If the test is run with `--verbose`, one can see the corresponding `curl` command to run:

```
hurl --verbose --test tests_ok/output.hurl
tests_ok/output.hurl: Running [1/1]
* Options:
...
* Request can be run with the following curl command:
* curl -H 'Content-Type: application/json' --data '{ "user": "bob" }' 'http://localhost:8000/output/endpoint1'
*
```
We want the test file to be run with `curl`. Running the test with `curl` in verbose mode should also shows that HTTP2 is not supported (Hurl is just "asserts on top of `curl`).

In the [tests suite](https://github.com/Orange-OpenSource/hurl/tree/master/integration/tests_ok), as we test on various platform (see https://github.com/Orange-OpenSource/hurl/actions/runs/3658000153 for instance), we just use `HTTP/*` assert for most test cases. On some case we are asserting `HTTP/1.0`and even on the same distribution we had some failure due to Flask upgrades (the app running integration tests).

That said, we can stress in the documentation that Hurl relies strongly on the system libcurl and its capacities (maybe in https://hurl.dev/docs/running-tests.html or https://hurl.dev/docs/frequently-asked-questions.html).

**Comment by dnsmichi** [comment]
From a user's perspective, it can be irritating when e.g. using different container images, thus getting different results. Or the Alpine image works differently than macOS native for example. It is not expected, and quite frankly said, I don't care about libcurl and its dependencies in the background, I only see `hurl` :) 

Documenting this behavior will help folks who run into the same problem as I did. I don't think everyone will catch the HTTP protocol version like I did, but the test case is there.

Noted that you can use wildcard pattern matching, I did not know that, thanks.

**Comment by jcamiel** [comment]
Fixes by #1217
